design/ErrorHandling.md

```diff
@@ -80,90 +79,244 @@ device.addEventListener('validationerror', (event) => {
+- The Promise could reject if the object is invalid. Con: Promises can only reject to DOMExceptions, meaning we have to express error categories via established DOMException error codes. (Also not sure if DOMException allows custom `message`s.)
 
-A recoverable error is exposed as a `GPUObjectStatusQuery`.
+### Operation Status
 
-```
-// (Exact form/type subject to change.)
-typedef Promise<GPUObjectStatus> GPUObjectStatusQuery;
+For the "Void-Returning" operations, errors are reported via `getLastOperationStatus()`.
+
+  - `GPUQueue.submit`
+  - `GPUQueue.signal`
+  - `GPUBuffer.setSubData`
+  - `GPUBuffer.unmap`
+  - `GPUBuffer.destroy`
+  - `GPUTexture.destroy`
+  - `GPUFence.getCompletedValue` (never fails)
+
+`device.getLastOperationStatus()` always resolves to the status of the most recent void-returning operation.
+If called multiple times in a row, it returns the same result (but not the same object).
+
+`getLastOperationStatus` returns a Promise that resolves to a `GPUOperationStatus` object describing the status of the last operation on this object.
+`GPUOperationStatus` includes a `type`, which describes the general category of error, as well as a `message` (an implementation-defined human-readable message string about the error).
+
+`getStatus` rejects if the device is lost.
+
+```webidl
+interface GPUOperationStatus {
+    readonly attribute boolean succeeded;
 
-typedef (GPUBuffer or GPUTexture) StatusableObject;
+    // These are only set if succeeded == false.
+    readonly attribute GPUErrorType? type;
+    readonly attribute DOMString? message;
+};
 
 partial interface GPUDevice {
-    GPUObjectStatusQuery getObjectStatus(StatusableObject object);
+    Promise<GPUOperationStatus> getLastOperationStatus();
 };
 ```
 
-A concrete example: When creating a buffer, the following logic applies:
+#### Alternatives
 
- - `createBuffer` returns a `GPUBuffer` object `buffer` immediately.
- - A `GPUObjectStatusQuery` can be obtained by calling `device.getObjectStatus(buffer)`.
-   At a later time, that query resolves to a `GPUObjectStatus` that is one of:
-    - Creation succeeded (`"valid"`).
-    - Creation encountered a recoverable error (`"out-of-memory"`).
-      (The application can then choose to retry a smaller allocation of a *new* `GPUBuffer`.)
-    - Creation encountered another type of error out of the control of the application (`"invalid"`).
+  - `{GPUQueue,GPUBuffer,GPUTexture}.getLastOperationStatus`: the same but per-object.
+    Cons: pollutes objects with rarely used methods; requires each object to keep data for its last operation status.
+    Pro: associated with the object that had the error.
 
-Regardless of any recovery efforts the application makes, if creation fails,
-the resulting object is invalid (and subject to error propagation).
+  - `getOperationStatuses`: returns a sequence of statuses for past operations.
+    Con: sequence will grow unboundedly unless there is some arbitrary limit.
```

Comment:
I think that's ok, in general we only care about whether at least one error happened or none happened.